src/starter.py

```diff
@@ -0,0 +1,211 @@
+    """
+    Run command and get its stdout if completed successfully
+    :param cmd: Command to run
+    :param cwd: Working directory
+    :return: Bytes from stdout
+    """
+    proc = await run(cmd, cwd, pipe_output=True)
+    out, err = await proc.communicate()
+    if proc.returncode != 0:
+        raise RuntimeError(f"{cmd} failed with code {proc.returncode}")
+    return out
+
+
+async def stop_process(proc: Process, timeout: float = 3.0) -> None:
+    """
+    Try to stop the process gracefully, otherwise killing it
+    :param proc: Process to stop
+    :param timeout: How long to wait for a graceful stop
+    """
+    if proc.returncode is not None:
+        return
+    proc.terminate()
+    await asyncio.wait_for(proc.wait(), timeout=timeout)
+    if proc.returncode is not None:
+        return
+    else:
+        proc.kill()
+
+
+async def setup_nginx() -> None:
+    """
+    Prepare Nginx.
+    Optionally setup HTTP Basic Auth, if environment variables are provided.
+    Process Nginx templates to build actual configuration files.
+    """
+    if user := os.environ.get("HTTP_BASIC_AUTH_USER"):
+        pwd = os.environ["HTTP_BASIC_AUTH_PASSWORD"]
+        await run_and_wait(["htpasswd", "-b", "-c", "/run/nginx/.htpasswd", user, pwd])
+        if not os.environ.get("HTTP_BASIC_AUTH_REALM"):
+            os.environ["HTTP_BASIC_AUTH_REALM"] = "Private area"
+
+    for f in ["/etc/nginx/conf.d/default.conf", "/etc/nginx/dataspace-headers.conf"]:
+        final_config = await get_output(["parse-template", f])
+        Path(f).write_bytes(final_config)
+        print(f"{f} is generated")
+
+        if os.environ.get("NGINX_DEBUG"):
+            print(f"=== {f} ===")
+            print(Path(f).read_text("utf8") + "\n")
+
+
+async def start_service(svc: Service, extra_env: Optional[dict] = None) -> Process:
+    """
+    Start a service.
+    :param svc: Service to run
+    :param extra_env: Extra environment variables
+    :return: Process
+    """
+    debug_msg(f"Starting service: {svc.cmd}")
+    return await run(svc.cmd, cwd=svc.cwd, extra_env=extra_env)
+
+
+async def wait_for_service(svc: Service, timeout: int):
+    """
+    Wait until a service and up and listening for either a port or socket.
+    Port takes precedence over a socket if both defined.
+    :param svc: Service to wait for
+    :param timeout: How long to wait in seconds
+    """
+    if not svc.socket and not svc.port:
+        raise RuntimeError("Either socket or port should be defined")
+
+    url = "http://localhost"
+    transport = httpx.AsyncHTTPTransport(uds=svc.socket)
+
+    if svc.port:
+        transport = httpx.AsyncHTTPTransport()
+        url += f":{svc.port}"
+
+    async with httpx.AsyncClient(transport=transport) as client:
+        wait_step = 0.3
+        max_attempts = int(timeout / wait_step)
+        for _ in range(max_attempts):
+            try:
+                await client.get(url)
```

Comment:
I believe you should add some timeout either to the request itself or to the client to ensure it won't hand in case the process starts, but is super-slow for some reason, or decides to hang after opening the connection before returning anything.